Effects: Add :nothrow_if_inbounds effect

Keno opened the pull request:
Review note: Against #43852 while that is still pending.

This adds a new effect `:nothrow_if_inbounds` that basically does what it says on the tin.
Running #43852 through its paces, I've found it highly effective, except in cases where
a function was only nothrow because of a boundscheck with unknown bounds. This PR
adds support to make use of `@inbounds` annotations to still let inlining remove calls
that are nothrow except for an unknown boundscheck.

simeonschaub commented on base/math.jl:
```diff
@@ -18,7 +18,11 @@ export sin, cos, sincos, tan, sinh, cosh, tanh, asin, acos, atan,
 import .Base: log, exp, sin, cos, tan, sinh, cosh, tanh, asin,
              acos, atan, asinh, acosh, atanh, sqrt, log2, log10,
              max, min, minmax, ^, exp2, muladd, rem,
+<<<<<<< HEAD
              exp10, expm1, log1p, @constprop
+=======
+             exp10, expm1, log1p, @assume_effects
+>>>>>>> bbdb280899... Mark ^(x::Float64, n::Integer) as locally terminating
```
Bad rebase?

Keno replied:
Yeah ...

oscardssmith commented:
Does this need anything other than a rebase?

Keno commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2301,6 +2319,35 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                     hd !== nothing && is_meta_expr_head(hd))
                 # these do not generate code
             else
+                if hd === :(=)
```
Bad rebase

KristofferC commented:
To save a bit of CI / notifications it might make sense to just check that it builds locally once before pushing.

Keno committed:
Effects: Add :nothrow_if_inbounds effect

This adds a new effect :nothrow_if_inbounds that basically does what it says on the tin.
Running #43852 through its paces, I've found it highly effective, except in cases where
a function was only nothrow because of a boundscheck with unknown bounds. This PR
adds support to make use of @inbounds annotations to still let inlining remove calls
that are nothrow except for an unknown boundscheck.

vtjnash commented:
Is this still in-progress or dead? There look like a lot of conflicting files

Keno commented:
I think this is dead. We changed the way that inbounds and effects interact, so we'd have to rethink this anyway.
